src/execution/index/art/art.cpp

```diff
@@ -988,67 +978,160 @@ void ART::CheckConstraintsForChunk(DataChunk &input, ConflictManager &conflict_m
+	auto vacuum_flags = InitializeVacuum(allocators);
+
+	// skip vacuum if no allocators require it
+	auto perform_vacuum = false;
+	for (const auto &vacuum_flag : vacuum_flags) {
+		if (vacuum_flag) {
+			perform_vacuum = true;
+			break;
+		}
+	}
+	if (!perform_vacuum) {
+		return;
+	}
+
+	// traverse the allocated memory of the tree to perform a vacuum
+	ARTNode::Vacuum(*this, tree, vacuum_flags);
+
+	// finalize the vacuum operation
+	FinalizeVacuum(allocators, vacuum_flags);
+	UpdateMemoryUsage();
+}
+
 //===--------------------------------------------------------------------===//
 // Merging
 //===--------------------------------------------------------------------===//
 
+vector<idx_t> ART::InitializeMerge(vector<FixedSizeAllocator *> &allocators) {
+
+	vector<idx_t> buffer_counts;
+	buffer_counts.reserve(allocators.size());
+	for (auto &allocator : allocators) {
+		buffer_counts.emplace_back(allocator->buffers.size());
+	}
+	return buffer_counts;
+}
+
 bool ART::MergeIndexes(IndexLock &state, Index *other_index) {
 
 	auto other_art = (ART *)other_index;
+	auto allocators = GetAllocators();
+	auto other_allocators = other_art->GetAllocators();
 
-	if (!this->tree) {
-		IncreaseMemorySize(other_art->memory_size);
-		tree = other_art->tree;
-		other_art->tree = nullptr;
-		return true;
+	if (tree) {
+		//  fully deserialize other_index, and traverse it to increment its buffer IDs
+		auto buffer_counts = InitializeMerge(allocators);
+		other_art->tree.InitializeMerge(*other_art, buffer_counts);
 	}
 
-	return Node::MergeARTs(this, other_art);
+	// merge the node storage
+	for (idx_t i = 0; i < allocators.size(); i++) {
+		allocators[i]->Merge(*other_allocators[i]);
+	}
+
+	// merge the ARTs
+	if (!tree.Merge(*this, other_art->tree)) {
+		return false;
+	}
+	return true;
 }
 
 //===--------------------------------------------------------------------===//
 // Utility
 //===--------------------------------------------------------------------===//
 
 string ART::ToString() {
+	string str = "[empty]";
 	if (tree) {
-		return tree->ToString(*this);
+		str = tree.ToString(*this);
 	}
-	return "[empty]";
+	UpdateMemoryUsage();
+	return str;
 }
 
-void ART::Verify() {
-#ifdef DEBUG
-	idx_t current_mem_size = 0;
-	if (tree) {
-		current_mem_size = tree->MemorySize(*this, true);
-	}
-	if (memory_size != current_mem_size) {
-		throw InternalException("Memory_size value (%d) does not match actual memory size (%d).", memory_size,
-		                        current_mem_size);
-	}
-#endif
+vector<FixedSizeAllocator *> ART::GetAllocators() const {
+
+	vector<FixedSizeAllocator *> allocators;
+
+	allocators.push_back(prefix_segments.get());
+	allocators.push_back(leaf_segments.get());
+	allocators.push_back(leaves.get());
+	allocators.push_back(n4_nodes.get());
+	allocators.push_back(n16_nodes.get());
+	allocators.push_back(n48_nodes.get());
+	allocators.push_back(n256_nodes.get());
+
+	return allocators;
```

Comment:
Good point - perhaps we can take this one step further and instead have a `vector<unique_ptr<FixedSizeAllocator>>` that holds all of the actual allocators, followed by methods to get specific ones (e.g. `FixedSizeAllocator &GetPrefixAllocator()`)

Reply:
I went with that solution initially, but then each allocation required vector access. The profiler complained about it, so I discarded it and went with the separate members. 

A lot has changed since then, though, and might be worth having a look again, the performance drawbacks could've been caused by something else, too.

Reply:
Instead of `GetPrefixAllocator()` I indexed the vector with the node type to get the respective allocator.